src/main/java/dk/kb/util/yaml/YAML.java

```diff
@@ -821,59 +841,68 @@ private void traverseList(List<String> yPath, Object yaml, MultipleValuesVisitor
      * The implementation supports conditional lookups as well. This means that a query for zoo.[foo=bar].baz returns
      * all values for baz, that are a child of zoo and has the sibling-key foo with the value bar.
      * @param yPath a list of path elements. This list contains all parts of a specified path, which is most likely
-     *              delivered through the {@link #visit(Object path, YAML yaml)}-method.
+     *              delivered through the {@link #visit(Object path, YAML yaml, YAMLVisitor visitor)}-method.
      * @param yaml the current place in the YAML file being traversed. This should be an instance of a Map.
      * @param visitor used to collect values that match the given path. The only function of this visitor is to collect
      *               values to its internal list of extracted values: {@link MultipleValuesVisitor#extractedValues}.
      */
-    private void traverseMap(List<String> yPath, Object yaml, MultipleValuesVisitor visitor) {
+    private void traverseMap(YPath yPath, Object yaml, MultipleValuesVisitor visitor) {
+        if (yPath.isEmpty()) {
+            return;
+        }
+
         // Quick fix cleaning entries as [foo=bar] to foo=bar.
-        removeBracketsFromPathElement(yPath);
+        YPath cleanedYPath = removeBracketsFromPathElement(yPath);
+        log.info(yPath.getFirst());
 
         Map<?, ?> map = (Map<?, ?>) yaml;
-        List<String> shortenedPath = removeTraversedEntry(yPath);
+        YPath shortenedPath = cleanedYPath.removeFirst();
 
-        Matcher conditionalMatch;
-        if (!yPath.isEmpty()){
-            conditionalMatch = ARRAY_CONDITIONAL.matcher(yPath.get(0));
-        } else {
-            conditionalMatch = ARRAY_CONDITIONAL.matcher("");
-        }
+        Matcher conditionalMatch = ARRAY_CONDITIONAL.matcher(cleanedYPath.getFirst());
 
-        if (!yPath.isEmpty() && yPath.get(0).equals("**") && yPath.size() > 1){
+        if (cleanedYPath.getFirst().equals("**") && cleanedYPath.size() > 1){
             for (Map.Entry<?, ?> entry : map.entrySet()) {
-                traverse(yPath, entry.getValue(), visitor);
+                traverse(cleanedYPath, entry.getValue(), visitor);
                 traverse(shortenedPath, entry.getValue(), visitor);
             }
-        } else if (!yPath.isEmpty() && yPath.get(0).equals("*")) {
+        } else if (cleanedYPath.getFirst().equals("**") && cleanedYPath.size() == 1) {
```

Comment:
Fully optional:

As you have made a class for handling `yPath`, consider adding methods mirroring the conceptual actions that are used a lot. If this case if could be
```Java
} else if (cleanedYPath.firstEquals("**") && cleanedYPath.isLast()) {
```
or maybe even collapsed to
```Java
} else if (cleanedYPath.singleEquals("**")) {
```
all depending on how many times the action is performed. It can easily be overdone with too many special methods that confuses more than it helps. e.g. I think `singleEquals` might be overdoing it.

Reply:
I've implemented `firstEquals` and `isLast`

Reply:
Please also add Javadoc.